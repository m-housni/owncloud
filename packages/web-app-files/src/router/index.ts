import { RouteConfig } from 'vue-router'

import {
  buildRoutes as buildCommonRoutes,
  isLocationCommonActive,
  createLocationCommon
} from './common'
import { buildRoutes as buildDeprecatedRoutes, isLocationActive } from './deprecated'
import {
  buildRoutes as buildPublicRoutes,
  createLocationPublic,
  isLocationPublicActive
} from './public'
import { RouteComponents } from './router'
import {
  buildRoutes as buildSharesRoutes,
  isLocationSharesActive,
  createLocationShares
} from './shares'
import {
  buildRoutes as buildSpacesRoutes,
  isLocationSpacesActive,
  createLocationSpaces
} from './spaces'
import {
  buildRoutes as buildTrashRoutes,
  isLocationTrashActive,
  createLocationTrash
} from './trash'
import { ActiveRouteDirectorFunc } from './utils'

const ROOT_ROUTE = {
  path: '/',
  redirect: (to) => createLocationSpaces('files-spaces-generic', to)
}

const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  ROOT_ROUTE,
  ...buildCommonRoutes(components),
  ...buildSharesRoutes(components),
  ...buildPublicRoutes(components),
  ...buildSpacesRoutes(components),
  ...buildTrashRoutes(components),
  ...buildDeprecatedRoutes()
]

export {
  createLocationCommon,
  createLocationShares,
  createLocationSpaces,
  createLocationPublic,
  isLocationCommonActive,
  isLocationSharesActive,
  isLocationSpacesActive,
  isLocationPublicActive,
  isLocationActive,
  isLocationTrashActive,
  createLocationTrash,
  buildRoutes,
  ActiveRouteDirectorFunc
}
