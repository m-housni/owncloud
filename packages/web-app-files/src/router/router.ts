import Vue, { ComponentOptions } from 'vue'

/**
 * we need to inject the vue files into the route builders,
 * this is because we also import the provided helpers from other js|ts files
 * like mixins, rollup seems to have a problem to import files which contain vue file imports
 * into js files which then again get imported by other vue files...
 */
export interface RouteComponents {
  App: ComponentOptions<Vue>
  Favorites: ComponentOptions<Vue>
  FilesDrop: ComponentOptions<Vue>
  PrivateLink: ComponentOptions<Vue>
  SearchResults: ComponentOptions<Vue>
  PublicLink: ComponentOptions<Vue>
  Shares: {
    SharedWithMe: ComponentOptions<Vue>
    SharedWithOthers: ComponentOptions<Vue>
    SharedViaLink: ComponentOptions<Vue>
  }
  Spaces: {
    DriveResolver: ComponentOptions<Vue>
    Projects: ComponentOptions<Vue>
  }
}
