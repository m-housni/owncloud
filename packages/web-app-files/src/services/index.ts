export * from './archiver'
export * from './thumbnail'
export { default as Registry } from './registry'
