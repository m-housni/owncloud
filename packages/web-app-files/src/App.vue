<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="oc-width-expand" />
  </main>
</template>
<script lang="ts">
import { defineComponent, watch } from 'vue'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { eventBus } from 'web-pkg/src/services/eventBus'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    watch(useRoute(), () => {
      store.dispatch('Files/resetFileSelection')
    })
  },
  data: () => ({
    dragareaEnabled: false
  }),
  created() {
    const dragOver = eventBus.subscribe('drag-over', this.onDragOver)
    const dragOut = eventBus.subscribe('drag-out', this.hideDropzone)
    const drop = eventBus.subscribe('drop', this.hideDropzone)

    this.$on('beforeDestroy', () => {
      eventBus.unsubscribe('drag-over', dragOver)
      eventBus.unsubscribe('drag-out', dragOut)
      eventBus.unsubscribe('drop', drop)
    })
  },

  methods: {
    hideDropzone() {
      this.dragareaEnabled = false
    },
    onDragOver(event) {
      this.dragareaEnabled = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
