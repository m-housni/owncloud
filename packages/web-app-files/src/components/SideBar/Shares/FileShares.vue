<template>
  <div id="oc-files-sharing-sidebar" class="oc-position-relative">
    <div class="oc-flex">
      <h3 v-translate class="oc-text-bold oc-text-medium oc-m-rm">Share with people</h3>
      <oc-contextual-helper
        v-if="helpersEnabled"
        class="oc-pl-xs"
        v-bind="inviteCollaboratorHelp"
      />
    </div>
    <invite-collaborator-form v-if="currentUserCanShare" key="new-collaborator" class="oc-my-s" />
    <p
      v-else
      key="no-reshare-permissions-message"
      data-testid="files-collaborators-no-reshare-permissions-message"
      v-text="noResharePermsMessage"
    />
    <div v-if="hasSharees" class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
      <h4 class="oc-text-bold oc-my-rm" v-text="sharedWithLabel" />
    </div>
    <template v-if="hasSharees">
      <ul
        id="files-collaborators-list"
        class="oc-list oc-list-divider oc-overflow-hidden"
        :class="{ 'oc-mb-l': showSpaceMembers, 'oc-m-rm': !showSpaceMembers }"
        :aria-label="$gettext('Share receivers')"
      >
        <li v-for="collaborator in displayCollaborators" :key="collaborator.key">
          <collaborator-list-item
            :share="collaborator"
            :modifiable="isShareModifiable(collaborator)"
            :shared-parent-route="getSharedParentRoute(collaborator)"
            @onDelete="$_ocCollaborators_deleteShare_trigger"
          />
        </li>
      </ul>
      <div v-if="showShareToggle" class="oc-flex oc-flex-center">
        <oc-button
          appearance="raw"
          @click="toggleShareesListCollapsed"
          v-text="collapseButtonTitle"
        />
      </div>
    </template>
    <template v-if="showSpaceMembers">
      <h4 class="oc-text-bold oc-my-s" v-text="spaceMemberLabel" />
      <ul
        id="space-collaborators-list"
        class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm"
        :aria-label="spaceMemberLabel"
      >
        <li v-for="collaborator in displaySpaceMembers" :key="collaborator.key">
          <collaborator-list-item :share="collaborator" :modifiable="false" />
        </li>
      </ul>
      <div v-if="showShareToggle" class="oc-flex oc-flex-center">
        <oc-button
          appearance="raw"
          @click="toggleShareesListCollapsed"
          v-text="collapseButtonTitle"
        />
      </div>
    </template>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions, mapState, mapMutations } from 'vuex'
import {
  useStore,
  useCapabilityProjectSpacesEnabled,
  useCapabilityShareJailEnabled,
  useCapabilityFilesSharingResharing
} from 'web-pkg/src/composables'
import { createLocationSpaces, isLocationSharesActive } from '../../../router'
import { textUtils } from '../../../helpers/textUtils'
import { getParentPaths } from '../../../helpers/path'
import { ShareTypes } from 'web-client/src/helpers/share'
import InviteCollaboratorForm from './Collaborators/InviteCollaborator/InviteCollaboratorForm.vue'
import CollaboratorListItem from './Collaborators/ListItem.vue'
import {
  shareInviteCollaboratorHelp,
  shareInviteCollaboratorHelpCern
} from '../../../helpers/contextualHelpers'
import { computed, defineComponent, PropType } from 'vue'
import { isProjectSpaceResource, SpaceResource } from 'web-client/src/helpers'
import { createFileRouteOptions } from 'web-pkg/src/helpers/router'

export default defineComponent({
  name: 'FileShares',
  components: {
    InviteCollaboratorForm,
    CollaboratorListItem
  },
  props: {
    space: {
      type: Object as PropType<SpaceResource>,
      required: false,
      default: null
    }
  },
  setup() {
    const store = useStore()
    const sharesListCollapsed = !store.getters.configuration.options.sidebar.shares.showAllOnLoad
    const currentUserIsMemberOfSpace = computed(() => {
      const userId = store.getters.user?.id
      if (!userId) {
        return false
      }
      return store.getters['runtime/spaces/spaceMembers'].some(
        (member) => member.collaborator?.name === userId
      )
    })

    return {
      sharesListCollapsed,
      currentUserIsMemberOfSpace,
      hasProjectSpaces: useCapabilityProjectSpacesEnabled(),
      hasShareJail: useCapabilityShareJailEnabled(),
      hasResharing: useCapabilityFilesSharingResharing()
    }
  },
  computed: {
    ...mapGetters('Files', ['highlightedFile', 'currentFileOutgoingCollaborators']),
    ...mapGetters(['configuration']),
    ...mapGetters('runtime/spaces', ['spaceMembers']),
    ...mapState('Files', ['incomingShares', 'sharesTree']),
    ...mapState(['user']),

    inviteCollaboratorHelp() {
      const cernFeatures = !!this.configuration?.options?.cernFeatures

      if (cernFeatures) {
        const mergedHelp = shareInviteCollaboratorHelp
        mergedHelp.list = [
          ...shareInviteCollaboratorHelpCern.list,
          ...shareInviteCollaboratorHelp.list
        ]
        return mergedHelp
      }

      return shareInviteCollaboratorHelp
    },
    helpersEnabled() {
      return this.configuration?.options?.contextHelpers
    },
    sharedWithLabel() {
      return this.$gettext('Shared with')
    },
    spaceMemberLabel() {
      return this.$gettext('Space members')
    },

    collapseButtonTitle() {
      return this.sharesListCollapsed ? this.$gettext('Show more') : this.$gettext('Show less')
    },

    hasSharees() {
      return this.collaborators.length > 0
    },

    collaborators() {
      return [...this.currentFileOutgoingCollaborators, ...this.indirectOutgoingShares]
        .sort(this.collaboratorsComparator)
        .map((collaborator) => {
          collaborator.key = 'collaborator-' + collaborator.id
          if (
            collaborator.owner.name !== collaborator.fileOwner.name &&
            collaborator.owner.name !== this.user.id
          ) {
            collaborator.resharers = [collaborator.owner]
          }
          return collaborator
        })
    },

    displayCollaborators() {
      if (this.collaborators.length > 3 && this.sharesListCollapsed) {
        return this.collaborators.slice(0, 3)
      }
      return this.collaborators
    },

    displaySpaceMembers() {
      if (this.spaceMembers.length > 3 && this.sharesListCollapsed) {
        return this.spaceMembers.slice(0, 3)
      }
      return this.spaceMembers
    },

    showShareToggle() {
      return this.spaceMembers.length > 3 || this.collaborators.length > 3
    },

    indirectOutgoingShares() {
      const allShares = []
      const parentPaths = getParentPaths(this.highlightedFile.path, false)
      if (parentPaths.length === 0) {
        return []
      }

      parentPaths.forEach((parentPath) => {
        const shares = this.sharesTree[parentPath]
        if (shares) {
          shares.forEach((share) => {
            if (share.outgoing && this.$_isCollaboratorShare(share)) {
              allShares.push({ ...share, key: 'indirect-collaborator-' + share.id })
            }
          })
        }
      })

      return allShares
    },

    currentUserCanShare() {
      if (this.highlightedFile.isReceivedShare() && !this.hasResharing) {
        return false
      }
      const isShareJail = this.space?.driveType === 'share'
      if (isShareJail && !this.hasResharing) {
        return false
      }

      return this.highlightedFile.canShare({ user: this.user })
    },
    noResharePermsMessage() {
      const translatedFile = this.$gettext("You don't have permission to share this file.")
      const translatedFolder = this.$gettext("You don't have permission to share this folder.")
      return this.highlightedFile.type === 'file' ? translatedFile : translatedFolder
    },
    showSpaceMembers() {
      return (
        this.space?.driveType === 'project' &&
        this.highlightedFile.type !== 'space' &&
        this.currentUserIsMemberOfSpace
      )
    }
  },
  methods: {
    ...mapActions('Files', ['deleteShare']),
    ...mapActions(['createModal', 'hideModal', 'showMessage']),
    ...mapMutations('Files', ['REMOVE_FILES']),

    toggleShareesListCollapsed() {
      this.sharesListCollapsed = !this.sharesListCollapsed
    },
    $_isCollaboratorShare(collaborator) {
      return ShareTypes.containsAnyValue(ShareTypes.authenticated, [collaborator.shareType])
    },
    collaboratorsComparator(c1, c2) {
      // Sorted by: type, direct, display name, creation date
      const c1DisplayName = c1.collaborator ? c1.collaborator.displayName : c1.displayName
      const c2DisplayName = c2.collaborator ? c2.collaborator.displayName : c2.displayName
      const name1 = c1DisplayName.toLowerCase().trim()
      const name2 = c2DisplayName.toLowerCase().trim()
      const c1UserShare = ShareTypes.containsAnyValue(ShareTypes.individuals, [c1.shareType])
      const c2UserShare = ShareTypes.containsAnyValue(ShareTypes.individuals, [c2.shareType])
      const c1DirectShare = !c1.indirect
      const c2DirectShare = !c2.indirect

      if (c1UserShare === c2UserShare) {
        if (c1DirectShare === c2DirectShare) {
          if (name1 === name2) {
            return textUtils.naturalSortCompare(c1.stime + '', c2.stime + '')
          }

          return textUtils.naturalSortCompare(name1, name2)
        }

        return c1DirectShare ? -1 : 1
      }

      return c1UserShare ? -1 : 1
    },

    $_ocCollaborators_deleteShare_trigger(share) {
      const modal = {
        variation: 'danger',
        icon: 'alarm-warning',
        title: this.$gettext('Remove share'),
        cancelText: this.$gettext('Cancel'),
        confirmText: this.$gettext('Remove'),
        message: this.$gettext('Are you sure you want to remove this share?'),
        hasInput: false,
        onCancel: this.hideModal,
        onConfirm: () => this.$_ocCollaborators_deleteShare(share)
      }

      this.createModal(modal)
    },

    async $_ocCollaborators_deleteShare(share) {
      let path = this.highlightedFile.path
      // sharing a share root from the share jail -> use resource name as path
      if (this.hasShareJail && path === '/') {
        path = `/${this.highlightedFile.name}`
      }

      const lastShareId =
        this.currentFileOutgoingCollaborators.length === 1
          ? this.currentFileOutgoingCollaborators[0].id
          : undefined

      try {
        await this.deleteShare({
          client: this.$client,
          share: share,
          path,
          storageId: this.highlightedFile.fileId,
          loadIndicators: !!lastShareId
        })

        this.hideModal()
        this.showMessage({
          title: this.$gettext('Share was removed successfully')
        })
        if (lastShareId && isLocationSharesActive(this.$router, 'files-shares-with-others')) {
          this.REMOVE_FILES([{ id: lastShareId }])
        }
      } catch (error) {
        console.error(error)
        this.showMessage({
          title: this.$gettext('Failed to remove share'),
          status: 'danger'
        })
      }
    },
    getSharedParentRoute(parentShare) {
      if (!parentShare.indirect) {
        return null
      }

      // TODO: this doesn't work on files-spaces-share routes?!
      if (this.space && this.sharesTree[parentShare.path]) {
        return createLocationSpaces(
          'files-spaces-generic',
          createFileRouteOptions(this.space, {
            path: parentShare.path,
            fileId: parentShare.file.source
          })
        )
      }

      return null
    },

    isShareModifiable(collaborator) {
      if (
        this.space &&
        isProjectSpaceResource(this.space) &&
        this.currentUserIsMemberOfSpace &&
        !this.space?.spaceRoles.manager.includes(this.user.uuid)
      ) {
        return false
      }

      return !collaborator.indirect
    }
  }
})
</script>

<style>
.avatars-wrapper {
  height: 40px;
}
</style>
