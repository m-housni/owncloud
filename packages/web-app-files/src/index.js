import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import translations from '../l10n/translations.json'
import quickActions from './quickActions'
import store from './store'
import { SDKSearch } from './search'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { archiverService, thumbnailService, Registry } from './services'
import fileSideBars from './fileSideBars'
import { buildRoutes } from './router'
import get from 'lodash-es/get'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
const navItems = [
  {
    name(capabilities) {
      return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
    },
    icon: appInfo.icon,
    route: {
      path: `/${appInfo.id}/spaces/personal`
    }
  },
  {
    name: $gettext('Favorites'),
    icon: 'star',
    route: {
      path: `/${appInfo.id}/favorites`
    },
    enabled(capabilities) {
      return capabilities.files && capabilities.files.favorites
    }
  },
  {
    name: $gettext('Shares'),
    icon: 'share-forward',
    route: {
      path: `/${appInfo.id}/shares`
    },
    activeFor: [{ path: `/${appInfo.id}/spaces/share` }],
    enabled(capabilities) {
      return capabilities.files_sharing?.api_enabled !== false
    }
  },
  {
    name: $gettext('Spaces'),
    icon: 'layout-grid',
    route: {
      path: `/${appInfo.id}/spaces/projects`
    },
    activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
    enabled(capabilities) {
      return capabilities.spaces && capabilities.spaces.projects === true
    }
  },
  {
    name: $gettext('Deleted files'),
    icon: 'delete-bin-5',
    route: {
      path: `/${appInfo.id}/trash`
    },
    enabled(capabilities) {
      return capabilities.dav && capabilities.dav.trashbin === '1.0'
    }
  }
]

export default {
  appInfo,
  store,
  routes: buildRoutes({
    App,
    Favorites,
    FilesDrop,
    SearchResults,
    Shares: {
      SharedViaLink,
      SharedWithMe,
      SharedWithOthers
    },
    Spaces: {
      DriveResolver: SpaceDriveResolver,
      Projects: SpaceProjects
    }
  }),
  navItems,
  quickActions,
  translations,
  ready({ router, store }) {
    Registry.sdkSearch = new SDKSearch(store, router)

    // when discussing the boot process of applications we need to implement a
    // registry that does not rely on call order, aka first register "on" and only after emit.
    eventBus.publish('app.search.register.provider', Registry.sdkSearch)

    archiverService.initialize(
      store.getters.configuration.server || window.location.origin,
      get(store, 'getters.capabilities.files.archivers', [
        {
          enabled: true,
          version: '1.0.0',
          formats: ['tar', 'zip'],
          archiver_url: `${store.getters.configuration.server}index.php/apps/files/ajax/download.php`
        }
      ])
    )
    // FIXME: Use capability data only as soon as available
    thumbnailService.initialize(
      get(store, 'getters.capabilities.files.thumbnail', {
        enabled: true,
        version: 'v0.1',
        supportedMimeTypes: store.getters.configuration?.options?.previewFileMimeTypes || []
      })
    )
  }
}
