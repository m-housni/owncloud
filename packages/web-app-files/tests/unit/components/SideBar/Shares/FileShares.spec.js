import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
import mockAxios from 'jest-mock-axios'
import GetTextPlugin from 'vue-gettext'
import Vuex from 'vuex'
import DesignSystem from '@ownclouders/design-system'
import stubs from '@/tests/unit/stubs'
import Users from '@/__fixtures__/users'
import Collaborators from '@/__fixtures__/collaborators'
import { spaceRoleManager } from 'web-client/src/helpers/share'
import * as reactivityComposables from 'web-pkg/src/composables/reactivity'
import * as routerComposables from 'web-pkg/src/composables/router'
import FileShares from 'web-app-files/src/components/SideBar/Shares/FileShares.vue'
import { buildSpace } from 'web-client/src/helpers'
import { clientService } from 'web-pkg/src/services'

const localVue = createLocalVue()
localVue.prototype.$clientService = clientService
localVue.use(DesignSystem)
localVue.use(Vuex)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

jest.mock('web-pkg/src/composables/reactivity')
jest.mock('web-pkg/src/composables/router')

const user = Users.alice
const collaborators = [Collaborators[0], Collaborators[1]]
const storageId = '1'

const selectors = {
  firstCollaboratorListItem: `div[data-testid="collaborator-user-item-${Collaborators[0].collaborator.name}"]`
}

describe('FileShares', () => {
  it('renders loading spinner while loading', () => {
    const wrapper = getShallowMountedWrapper(
      {
        user
      },
      true
    )
    expect(wrapper).toMatchSnapshot()
  })

  describe('if currentUser can share', () => {
    it('initially renders add people dialog', () => {
      const wrapper = getShallowMountedWrapper({
        user
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if currentUser can not share', () => {
    it('initially renders no share permission message', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        canShare: false
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if there are no collaborators', () => {
    it('does not render avatar wrapper or collaborator list', () => {
      const wrapper = getShallowMountedWrapper({
        user
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if there are collaborators present', () => {
    it('renders sharedWithLabel and sharee list', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        outgoingCollaborators: collaborators
      })
      expect(wrapper).toMatchSnapshot()
    })

    it('reacts on delete events by collaborator list items', async () => {
      const spyOnCollaboratorDeleteTrigger = jest
        .spyOn(FileShares.methods, '$_ocCollaborators_deleteShare_trigger')
        .mockImplementation()
      const wrapper = getMountedWrapper({
        user,
        outgoingCollaborators: collaborators
      })
      wrapper.find(selectors.firstCollaboratorListItem).vm.$emit('onDelete')
      await wrapper.vm.$nextTick()
      expect(spyOnCollaboratorDeleteTrigger).toHaveBeenCalledTimes(1)
    })
    it('correctly passes the shared parent route to the collaborator list item', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        outgoingCollaborators: [{ ...Collaborators[0], indirect: true }]
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('current space', () => {
    afterEach(() => {
      mockAxios.reset()
    })
    it('loads space members if a space is given', () => {
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: { role: spaceRoleManager.name }
        })
      })

      const spaceMock = buildSpace({
        id: '1',
        driveType: 'project',
        root: {
          permissions: [
            {
              roles: ['manager'],
              grantedTo: [{ user: { id: user.uuid } }]
            }
          ]
        }
      })
      const wrapper = getShallowMountedWrapper({
        user,
        spaces: [spaceMock],
        spaceMembers: [{ id: 1 }],
        currentUserIsMemberOfSpace: true
      })
      expect(wrapper.find('#space-collaborators-list').exists()).toBeTruthy()
    })
    it('does not load space members if no space is given', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        spaceMembers: [{ id: 1 }]
      })

      expect(wrapper.find('#space-collaborators-list').exists()).toBeFalsy()
    })
  })
})

function getResource({
  filename = 'testFile',
  extension = 'txt',
  type = 'file',
  owner = 'user0',
  canShare = true
}) {
  return {
    id: '4',
    fileId: '4',
    icon: type,
    name: type === 'file' ? `${filename}.${extension}` : filename,
    extension: extension,
    path: type === 'file' ? `/${filename}.${extension}` : `/${filename}`,
    type,
    isFolder: type === 'folder',
    mdate: 'Mon, 12 Jul 2021 11:04:33 GMT',
    size: '163',
    indicators: [],
    permissions: 'RDNVW',
    starred: false,
    etag: '"89128c0e8122002db57bd19c9ec33004"',
    shareTypes: [],
    downloadURL: '',
    ownerDisplayName: owner.displayname,
    ownerId: owner.id,
    canDownload: () => true,
    isReceivedShare: () => true,
    canBeDeleted: () => true,
    canRename: () => true,
    canShare: () => canShare,
    canDeny: () => false
  }
}

const storeOptions = (data) => {
  const {
    user,
    outgoingCollaborators = [],
    incomingCollaborators = [],
    canShare = true,
    spaces = [],
    spaceMembers = []
  } = data
  return {
    actions: {
      createModal: jest.fn(),
      hideModal: jest.fn(),
      showMessage: jest.fn()
    },
    state: {
      user
    },
    modules: {
      Files: {
        state: {
          incomingShares: incomingCollaborators,
          sharesTree: { [Collaborators[0].path]: [Collaborators[0]] }
        },
        namespaced: true,
        getters: {
          highlightedFile: () => {
            return getResource({ filename: 'testfile', extension: 'jpg', type: 'file', canShare })
          },
          currentFileOutgoingCollaborators: () => outgoingCollaborators,
          sharesTreeLoading: () => false
        },
        actions: {
          loadSharesTree: jest.fn(),
          deleteShare: jest.fn()
        },
        mutations: {
          SET_HIGHLIGHTED_FILE(state, file) {
            const newFile = getResource(file)
            state.highlightedFile = newFile
          }
        }
      },
      runtime: {
        namespaced: true,
        modules: {
          auth: {
            namespaced: true,
            getters: {
              accessToken: () => 'GFwHKXdsMgoFwt'
            }
          },
          spaces: {
            namespaced: true,
            state: { spaces },
            getters: {
              spaceMembers: () => spaceMembers
            }
          }
        }
      }
    },
    getters: {
      configuration: jest.fn(() => ({
        options: {
          sidebar: {
            shares: {
              showAllOnLoad: false
            }
          }
        },
        server: 'http://example.com/'
      })),
      user: () => user,
      capabilities: () => {
        return {
          files_sharing: {
            user: {
              expire_date: {
                enabled: true,
                days: 10
              }
            },
            group: {
              expire_date: {
                enabled: true,
                days: 10
              }
            }
          }
        }
      }
    }
  }
}

function getMountedWrapper(data) {
  routerComposables.useRouteParam.mockReturnValue(() => storageId)

  return mount(FileShares, {
    localVue,
    provide: {
      incomingParentShare: {}
    },
    setup: () => ({
      currentStorageId: storageId
    }),
    store: createStore(data),
    stubs: {
      ...stubs,
      'oc-button': false
    },
    mocks: {
      $router: {
        currentRoute: { name: 'some-route' },
        resolve: (r) => {
          return {
            href: r.name
          }
        }
      }
    }
  })
}

function getShallowMountedWrapper(data, loading = false) {
  reactivityComposables.useDebouncedRef.mockImplementationOnce(() => loading)
  routerComposables.useRouteParam.mockReturnValue(() => storageId)
  const { spaces = [], currentUserIsMemberOfSpace } = data

  return shallowMount(FileShares, {
    localVue,
    setup: () => ({
      currentStorageId: storageId,
      hasResharing: false,
      currentUserIsMemberOfSpace
    }),
    provide: {
      incomingParentShare: {}
    },
    store: createStore(data),
    stubs: {
      ...stubs,
      'oc-button': true
    },
    propsData: {
      space: spaces.length ? spaces[0] : null
    },
    mocks: {
      $router: {
        currentRoute: { name: 'some-route' },
        resolve: (r) => {
          return {
            href: r.name
          }
        }
      }
    }
  })
}

function createStore(data) {
  return new Vuex.Store(storeOptions(data))
}
