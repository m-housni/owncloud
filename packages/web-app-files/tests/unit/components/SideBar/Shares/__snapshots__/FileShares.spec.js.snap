// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileShares if currentUser can not share initially renders no share permission message 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Share with people" data-current-language="en_US">Share with people</h3>
    <!---->
  </div>
  <p data-testid="files-collaborators-no-reshare-permissions-message">You don't have permission to share this file.</p>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`FileShares if currentUser can share initially renders add people dialog 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Share with people" data-current-language="en_US">Share with people</h3>
    <!---->
  </div>
  <p data-testid="files-collaborators-no-reshare-permissions-message">You don't have permission to share this file.</p>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`FileShares if there are collaborators present correctly passes the shared parent route to the collaborator list item 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Share with people" data-current-language="en_US">Share with people</h3>
    <!---->
  </div>
  <p data-testid="files-collaborators-no-reshare-permissions-message">You don't have permission to share this file.</p>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-bold oc-my-rm">Shared with</h4>
  </div>
  <ul id="files-collaborators-list" aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
  <!---->
  <!---->
</div>
`;

exports[`FileShares if there are collaborators present renders sharedWithLabel and sharee list 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Share with people" data-current-language="en_US">Share with people</h3>
    <!---->
  </div>
  <p data-testid="files-collaborators-no-reshare-permissions-message">You don't have permission to share this file.</p>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-bold oc-my-rm">Shared with</h4>
  </div>
  <ul id="files-collaborators-list" aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
  </ul>
  <!---->
  <!---->
</div>
`;

exports[`FileShares if there are no collaborators does not render avatar wrapper or collaborator list 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Share with people" data-current-language="en_US">Share with people</h3>
    <!---->
  </div>
  <p data-testid="files-collaborators-no-reshare-permissions-message">You don't have permission to share this file.</p>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`FileShares renders loading spinner while loading 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <div class="oc-flex">
    <h3 class="oc-text-bold oc-text-medium oc-m-rm" data-msgid="Share with people" data-current-language="en_US">Share with people</h3>
    <!---->
  </div>
  <p data-testid="files-collaborators-no-reshare-permissions-message">You don't have permission to share this file.</p>
  <!---->
  <!---->
  <!---->
</div>
`;
